lib/scholar/metrics.ex

```diff
@@ -331,6 +331,44 @@ defmodule Scholar.Metrics do
     Nx.divide(true_negatives, false_positives + true_negatives + 1.0e-16)
   end
 
+  @doc ~S"""
+  Calculates the confusion matrix.
+
+  ## Argument Shapes
+
+    * `y_true` - $\(d_0, d_1, ..., d_n\)$
+    * `y_pred` - $\(d_0, d_1, ..., d_n\)$
+
+  ## Examples
+
+      iex> y_true = Nx.tensor([0, 0, 1, 1, 2, 2], type: {:u, 32})
+      iex> y_pred = Nx.tensor([0, 1, 0, 2, 2, 2], type: {:u, 32})
+      iex> Scholar.Metrics.confusion_matrix(y_true, y_pred, num_classes: 3)
+      #Nx.Tensor<
+        s64[3][3]
+        [
+          [1, 1, 0],
+          [1, 0, 1],
+          [0, 0, 2]
+        ]
+      >
+  """
+  defn confusion_matrix(y_true, y_pred, opts \\ []) do
+    num_classes = opts[:num_classes]
+
+    indices =
+      Nx.iota({Nx.size(y_true)})
+      |> then(&[Nx.take(y_true, &1), Nx.take(y_pred, &1)])
+      |> Nx.stack()
+      |> Nx.transpose()
```

Comment:
Please see my comment on Nx.new_axis + Nx.concatenate

Nx.transpose is costly and if you concatenate along the correct axis beforehand it will be unnecessary.

Torchx.backend uses this quite frequently, IIRC

Reply:
Sorry I didn't see that before the previous commit. 😅 I think it should be okay now (except the backend thing), I wonder if there is anything else to improve?

Reply:
I think it's just the backend thing. I'm not at the PC right now, but if you cant find how to get it and pass it along, I can deal with this later :)

Reply:
I was looking at the docs wishing to find something like `Nx.tensor_backend/1` that gives me the backend of a tensor, but I didn't find one that looks like the correct function...I guess I'll leave that here, sorry.

Reply:
let's leave this as a TO-DO then. Worst case scenario the defn will use the default backend with default options to allocate the new tensors and that's probably the same env in which other tensors live anyway